packages/ptokens-assets-evm/src/lib/index.ts

```diff
@@ -1,183 +1,159 @@
+export const getEventPayload = (_log: Log): `0x${string}` => {
+  const decodedLogArgs = decodeAdapterLog(_log)
+  if (!isSwapLog(decodedLogArgs)) throw new Error('Invalid swap event log format')
+  // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
+  return concat([pad(_log.address), sha256(concat(_log.topics), 'hex'), decodedLogArgs.eventBytes.content])
+}
+
+export const getEventPreImage = (_log: Log, _context: `0x${string}`): `0x${string}` => {
+  if (!_log.blockHash) throw new Error('Missing block hash')
+  if (!_log.transactionHash) throw new Error('Missing transaction hash')
+  return concat([_context, _log.blockHash, _log.transactionHash, getEventPayload(_log)])
 }
 
-export const getOperationIdFromLog = (_log: Log, _networkId: NetworkId | null = null) => {
+export const getSwapEventId = (_log: Log, _context: `0x${string}`): `0x${string}` =>
+  sha256(getEventPreImage(_log, _context), 'hex')
+
+export const getSettleEventId = (_log: Log, _context: `0x${string}`): `0x${string}` =>
+  sha256(getEventPreImage(_log, _context), 'hex') // is this correct?
+
+export const serializeOperation = (operation: Operation) => [
+  operation.blockId,
+  operation.txId,
+  operation.nonce,
+  operation.erc20,
+  operation.originChainId,
+  operation.destinationChainId,
+  operation.amount,
+  operation.sender,
+  operation.recipient,
+  operation.data,
+]
+
+export const isSwapLog = (
+  args: ReturnType<typeof decodeAdapterLog>,
+): args is { nonce: bigint; eventBytes: { content: `0x${string}` } } =>
+  typeof args === 'object' && args !== null && 'nonce' in args && 'eventBytes' in args
+
+export const isSettleLog = (args: ReturnType<typeof decodeAdapterLog>): args is { eventId: `0x${string}` } =>
+  typeof args === 'object' && args !== null && 'eventId' in args
+
+export const decodeAdapterLog = (_log: Log) => {
   const decodedLog = decodeEventLog({
-    abi: pNetworkHubAbi,
+    abi: pNetworkAdapterAbi,
     ..._log,
   })
-  return getOperationIdFromObj(
-    Object.assign(
-      {},
-      'operation' in decodedLog.args ? decodedLog.args.operation : decodedLog.args,
-      {
-        transactionHash: _log.transactionHash,
-        blockHash: _log.blockHash,
-      },
-      _networkId ? { networkId: _networkId } : {},
-    ),
+  // eslint-disable-next-line @typescript-eslint/no-unsafe-return
+  return decodedLog.args
+}
+
+export const getOperationFromLog = (_log: Log, _chainId: number): Operation => {
+  const SIX_WORDS_LENGTH = 32 * 6
+
+  if (!_log.blockHash) throw new Error('blockHash not retreived correctly')
+  if (!_log.transactionHash) throw new Error('transactionHash not retreived correctly')
+
+  const decodedLogArgs = decodeAdapterLog(_log)
+  if (!isSwapLog(decodedLogArgs)) throw new Error('Invalid swap event log format')
+  const { nonce, eventBytes } = decodedLogArgs
+
+  const [decodedNonce, erc20, destinationChainId, amountTransferred, sender, hexRecipientLength] = decodeAbiParameters(
```

Comment:
I don't think this is correct, since the bytes from content are not encoded using the `abi.encoding`. They are just concatenated one after the other, so a shifted `slice` over the content should work. Like what is being done [here](https://github.com/pnetwork-association/pnetwork-v4/blob/master/solidity/src/PAM.sol#L143-L150)

Reply:
This works because they are exactly 32 bytes. In any case maybe it is confusing, I can use slice.